File: platform/backend/app/schemas/deployment.py
"""Pydantic schemas para deployments."""

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


class WorkspaceAdvancedConfig(BaseModel):
    """Config opcional do workspace exposta na aba "Avançado" do wizard.

    Campos default-derivados — usuario casual nao precisa preencher nada.
    Cluster sizing respeita driver/worker separados, autoscale opcional,
    policy enforcement e custom tags pra billing tracking.
    """

    root_bucket: str | None = None
    network_cidr: str | None = None
    admin_email: str | None = None
    metastore_id: str | None = None
    # Cluster identity
    cluster_name: str | None = None
    # Cluster sizing — node_type aplica a worker; driver_node_type opcional
    # (None = mesmo tipo do worker)
    cluster_node_type: str | None = None
    cluster_driver_node_type: str | None = None
    cluster_num_workers: int | None = None
    cluster_spark_version: str | None = None
    # Autoscale (se min/max set, sobrepõe num_workers)
    cluster_autoscale_min: int | None = None
    cluster_autoscale_max: int | None = None
    cluster_autotermination_min: int | None = None
    # Policy enforcement (force allowlist + max workers + ttl)
    cluster_policy_id: str | None = None
    # Policy custom JSON — saga registra no workspace + atrela ao cluster.
    # Mutuamente exclusivo com cluster_policy_id (custom tem precedencia).
    cluster_policy_definition: str | None = None
    # Custom tags propagadas pra AWS billing + Databricks usage
    cluster_tags: dict[str, str] | None = None
    # Observer Agent LLM override (per-pipeline). Sem isso, usa default
    # da empresa (company.preferred_provider/model).
    observer_llm_provider: str | None = None
    observer_llm_model: str | None = None


class DeploymentConfigIn(BaseModel):
    name: str
    environment: Literal["dev", "staging", "prod"] = "prod"
    tags: dict[str, str] = Field(default_factory=dict)
    credentials: dict[str, str] = Field(default_factory=dict)
    env_vars: dict[str, str] = Field(default_factory=dict)
    # Workspace selection: existing (skip provisioning) ou new (full saga).
    # Se ausente, default = "new" pra preservar compat com wizard antigo.
    workspace_mode: Literal["existing", "new"] = "new"
    # Modo existing: id do workspace alvo (Databricks Account API)
    workspace_id: str | None = None
    # Modo new: nome customizado do workspace (override do auto company-suffix)
    workspace_name: str | None = None
    advanced: WorkspaceAdvancedConfig | None = None


class DeploymentCreateRequest(BaseModel):
    template_slug: str
    config: DeploymentConfigIn


class DeploymentStepResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    step_id: str
    name: str
    description: str | None
    status: str
    order_index: int
    started_at: datetime | None
    finished_at: datetime | None
    duration_ms: int | None
    error_message: str | None


class DeploymentLogResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    level: str
    message: str
    step_id: str | None
    created_at: datetime


class DeploymentResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    template_slug: str
    template_name: str
    name: str
    environment: str
    config: dict
    status: str
    created_at: datetime
    started_at: datetime | None
    finished_at: datetime | None
    duration_ms: int | None
    pipeline_id: str | None
    steps: list[DeploymentStepResponse] = Field(default_factory=list)
    logs: list[DeploymentLogResponse] = Field(default_factory=list)


class DeploymentListItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    template_slug: str
    template_name: str
    name: str
    environment: str
    status: str
    created_at: datetime
    finished_at: datetime | None
    duration_ms: int | None


class DeploymentEvent(BaseModel):
    """Payload emitido via SSE para um deployment em execucao."""

    type: Literal["step_update", "log", "status_change", "complete", "error"]
    deployment_id: str
    data: dict

File: platform/frontend/app/types/deployment.ts
export type DeploymentStatus = "pending" | "running" | "success" | "failed" | "cancelled"
export type StepStatus = "pending" | "running" | "success" | "failed" | "skipped"
export type LogLevel = "info" | "warn" | "error" | "success" | "debug"

export interface SagaStepState {
  id: string
  name: string
  description?: string
  status: StepStatus
  startedAt?: string
  finishedAt?: string
  durationMs?: number
  errorMessage?: string
}

export interface LogEntry {
  id: string
  timestamp: string
  level: LogLevel
  message: string
  step?: string
}

export interface WorkspaceAdvancedConfig {
  rootBucket?: string
  networkCidr?: string
  adminEmail?: string
  metastoreId?: string
  /** Cluster reuse strategy: "existing" usa clusterId, "new" cria com nodeType + workers. */
  clusterMode?: "existing" | "new"
  /** Compute model: ephemeral (Job Compute, ~1/3 DBU, sem custo idle) ou persistent (all-purpose). Default ephemeral. */
  clusterCompute?: "ephemeral" | "persistent"
  /** existing mode: id do cluster ja running no workspace. */
  clusterId?: string
  /** Custom name; default = "medallion-pipeline". */
  clusterName?: string
  /** Worker node type AWS (ex m5d.large, m5d.xlarge). Default = m5d.large. */
  clusterNodeType?: string
  /** Driver node type AWS — separado pra otimizar custos. Default = mesmo do worker. */
  clusterDriverNodeType?: string
  /** Numero de workers fixos. Ignorado se autoscale set. Default = 2. */
  clusterNumWorkers?: number
  /** Autoscale min workers. Sobrepoe num_workers. */
  clusterAutoscaleMin?: number
  /** Autoscale max workers. Sobrepoe num_workers. */
  clusterAutoscaleMax?: number
  /** Auto-terminate idle minutes. Default = 30. */
  clusterAutoterminationMin?: number
  /** Versao Databricks Runtime (ex 15.4.x-scala2.12). Default = 15.4 LTS. */
  clusterSparkVersion?: string
  /** Cluster policy id (force allowlist + max workers + ttl). */
  clusterPolicyId?: string
  /** Cluster policy JSON definition — saga registra no workspace. */
  clusterPolicyDefinition?: string
  /** Custom tags propagadas pra AWS billing + Databricks usage. */
  clusterTags?: Record<string, string>
  /** Observer Agent LLM provider (anthropic/openai/google). Vazio = default empresa. */
  observerLlmProvider?: string
  /** Observer Agent model id literal (ex claude-opus-4-7). Vazio = default empresa. */
  observerLlmModel?: string
}

/** Credenciais conhecidas pre-preenchidas no wizard. Outras keys aceitas via index. */
export interface DeploymentCredentialsMap {
  aws_access_key_id: string
  aws_secret_access_key: string
  aws_region: string
  databricks_host: string
  databricks_token: string
  github_token: string
  [key: string]: string
}

export interface DeploymentConfig {
  name: string
  environment: "dev" | "staging" | "prod"
  tags: Record<string, string>
  credentials: DeploymentCredentialsMap
  envVars: Record<string, string>
  /** "existing" reusa workspace alvo (skip provisioning). "new" cria do zero. */
  workspaceMode?: "existing" | "new"
  /** workspaceMode=existing: id do workspace alvo (Databricks Account API). */
  workspaceId?: string
  /** workspaceMode=new: nome customizado do workspace (override do auto). */
  workspaceName?: string
  advanced?: WorkspaceAdvancedConfig
}

export interface Deployment {
  id: string
  templateSlug: string
  templateName: string
  config: DeploymentConfig
  status: DeploymentStatus
  createdAt: string
  startedAt?: string
  finishedAt?: string
  durationMs?: number
  steps: SagaStepState[]
  logs: LogEntry[]
  createdBy: string
  pipelineId?: string
}
